### ISSUE TITLE
Validation and sanitization API friction

### ISSUE THREAD

**Issue by melkio** [Particular/NServiceBus.AzureServiceBus#252]
Sanitization strategies provided by transport (`EndpointOrientedTopologySanitization` and `ThrowOnFailingSanitization`) have a dependency on configured `IValidationStrategy`.

Sanitization strategies use configured validation strategy to evaluate if provided value is valid or not to drive sanitization behavior. This check is useless since `AddressingLogic` [invokes sanitization only if validation fails](https://github.com/Particular/NServiceBus.AzureServiceBus/blob/release-7.0.0/src/Transport/Addressing/AddressingLogic.cs#L20)

### Plan of Attack

- [x] Change API 
- [x] Update documentation  => [PR #1629](https://github.com/Particular/docs.particular.net/pull/1629)
  - [x] API doco + existing RC issue Particular/docs.particular.net#1262
  - [x] Upgrade guide
  - [x] Samples
- End To End tests
  - [x] Transports compatability tests [here](https://github.com/Particular/EndToEnd/blob/master/src/TransportCompatibilityTests/AzureServiceBusV7/EndpointFacade.cs#L32)
  - [x] Performance tests [here](https://github.com/Particular/EndToEnd/blob/master/src/PerformanceTests/Transport.V6.AzureServiceBus.V7/AzureServiceBusProfile.cs)
- [x] Publish new Beta7 to nuget

**Comment by SeanFeldman** [comment]
> This check is useless since AddressingLogic invokes sanitization only if validation fails

Not quite. Addressing logic is:

```c#
public string Apply(string entityname, EntityType entityType)
{
	var path = composition.GetEntityPath(entityname, entityType);
	var validationResult = validationStrategy.IsValid(path, entityType); 
	if (!validationResult.IsValid)
	{
		path = sanitizationStrategy.Sanitize(path, entityType);
	}
	return path;
}
```

For `ThrowOnFailingSanitization` strategy we could drop the dependency on the validation strategy. But if we'd do that, we will be loosing the guidance in the error message in the exception thrown to the users. We _could_ pass in `validationResult` from the `AddressingLogic` rather than depending on a validation strategy.

Assume `validationResult.IsValid` is `false when `EndpointOrientedTopologySanitization` is running`. Here's the catch: the strategy doesn't know _what_ was invalid. Looking at `EndpointOrientedTopologySanitization`:

```c#
public string Sanitize(string entityPathOrName, EntityType entityType)
{
    // ...
	// remove invalid characters
	var rgx = new Regex(@"[^a-zA-Z0-9\-\._]");
	entityPathOrName = rgx.Replace(entityPathOrName, ""); // <== sanitization of characters is conditional, not absolute

	var validationResult = validationStrategy.IsValid(entityPathOrName, entityType);
	if (!validationResult.IsValid) // <== sanitization of the length is conditional, not absolute
	{
		// turn long name into a guid
		entityPathOrName = MD5DeterministicNameBuilder.Build(entityPathOrName);
	}

	return entityPathOrName;
}
```
While the sanitization strategy knows it **has** to remove the invalid characters, it doesn't know if it **should or should not** hash the path/name.  Again, if we'd pass `validationResult` into this sanitization strategy (just like I propose for `ThrowOnFailingSanitization` above), and add a  more granular option of reporting what's valid, we'd have the following `ValidationResult`: 

```c#
class ValidationResult
{
  public bool CharactersAreValid => ...;
  public bool LengthIsValid => ...;
  public bool IsValid => CharactersAreValid && LengthIsValid;
}
```

Which would allow `AddressingLogic` to flow the validation result from validation strategy into sanitization strategy in the following manner:

```c#
public string Apply(string entityname, EntityType entityType)
{
	var path = composition.GetEntityPath(entityname, entityType);
	var validationResult = validationStrategy.IsValid(path, entityType); 
	if (!validationResult.IsValid)
	{
		path = sanitizationStrategy.Sanitize(path, entityType, validationResult); // flow the context
	}
	return path;
}
```

and sanitization API would be 
`string Sanitize(string entityPathOrName, EntityType entityType, ValidationResult validationResult);`

Making sanitization strategies perform sanitization solely (`EndpointOrientedTopologySanitization` example):

 ```c#
public string Sanitize(string entityPathOrName, EntityType entityType, ValidationResult validationResult)
{
    // ...
	// remove invalid characters
	var rgx = new Regex(@"[^a-zA-Z0-9\-\._]");
	entityPathOrName = rgx.Replace(entityPathOrName, ""); // <== sanitization of characters is conditional, not absolute

	if (!validationResult.LengthIsValid) // <== sanitization of the length is conditional, not absolute
	{
		// turn long name into a guid
		entityPathOrName = MD5DeterministicNameBuilder.Build(entityPathOrName);
	}

	return entityPathOrName;
}
```

(`ThrowOnFailingSanitization`  example):

```c#
public string Sanitize(string entityPathOrName, EntityType entityType, ValidationResult validationResult)
{
	var pathOrName = "path";

	if (entityType == EntityType.Rule || entityType == EntityType.Subscription)
	{
		pathOrName = "name";
	}

	if (!validationResult.IsValid)
	{
		throw new Exception($"Invalid {entityType} entity {pathOrName} `{entityPathOrName}` that cannot be used with Azure Service Bus. Errors: " + string.Join("; ", validationResult.Errors) 
							+ Environment.NewLine + "Use `Sanitization().UseStrategy<ISanitizationStrategy>()` configuration API to register a custom sanitization strategy if required.");
	}

	return entityPathOrName;
}
```

**Comment by SeanFeldman** [comment]
TL;DR version:
Validation execution during sanitization is not _needed_, but validation result is _required_.

We can remove validation from sanitization if we 

1. provide a more granualr validation result (invalid characters and lenth violation)
1. flow validation results into sanitization

@Particular/azure-service-bus-maintainers thoughts?

**Comment by SeanFeldman** [comment]
@melkio I've put a spike for the lengthy comment of mine (in hindsight, perhaps should have done spike only).

**Comment by SeanFeldman** [comment]
After discussion with @melkio there's a more basic question I'd like to be answered. It will have an impact on how `IValidationStrategy` and `ISanitizationStrategy` could change.

_Should these two strategies be merged into one as they are both required for a **entity path/name** validation & sanitization?_

The first spike (#255) has removed the need to injecting validation strategy into sanitization strategy. Instead, it flows `ValidationResult` into sanitization as a context parameter when sanitizing an entity path/name. This still has somewhat of a dependency between validation _result_ and sanitization. We agreed that the focus of the two will solely remain characters and length, so it's not the issue of the fine-grained validation result, but the fact that it exists only to bridge from validation to sanitization. I.e. sanitization on its own w/o appropriate validation would be an awkward scenario.

@Particular/azure-service-bus-maintainers do you agree we should review the concept of 2 strategies and see if it makes sense? (2nd spike)

**Comment by SeanFeldman** [comment]
Update:
**SPIKE #255 - Do not perform validation in sanitization code**
`ValidationResult` created by validation is passed into sanitization

**SPIKE #257 - Validation and sanitization combined into ISanitizationStrategy**
Validation and sanitization is one strategy

**[SPIKE] #260 - Sanitization strategy per entity type**
Validation and sanitization is one strategy. Each `EntityType` can have it's own sanitization strategy.
Defaults changed: `ForwardingTopology` is using `ThrowOnFailingSanitizationStrategyFor[entitytype]` and `EndpointOrientedTopology` is using `EndpointOrientedTopologySanitizationFor[entitytype]`.

**Comment by yvesgoeleven** [comment]
I don't like spike #257 and #260. For the reason that it makes it hard for users to inject their existing naming conventions. Right now users can set their naming rules easily by means of a publicely available func, IValidationStrategy was intended as wrapper for this func. With IValidationStrategy & ISanitizationStrategy thrown together it will become hard for them to just set their naming rules as they'd have to implement a full sanitization strategy (or set of strategies). So my vote goes to #255

**Comment by SeanFeldman** [comment]
I'll bring one more disadvantageous point for combining validation and sanitization - order matters. Character sanitization can cancel length sanitization, which is undesired.

**Comment by melkio** [comment]
Also in the current release oreder matters, isn't it? Transport sanitizes chars before validating again to check if length is fine after the first step of sanitization.

Could you take a look to this [comment](https://github.com/Particular/NServiceBus.AzureServiceBus/pull/255/files#r66956187) related to "ordering" to validate what I wrote?

**Comment by SeanFeldman** [comment]
@melkio correct. Order matters in the current version.

**Comment by SeanFeldman** [comment]
#### Summary of the ASB Maintainers call from today

ASB maintainers focused on this issue as it raised some issues associated with the spikes and revealed a deeper issue - [ability to validate & sanitise individual entity types](https://github.com/Particular/NServiceBus.AzureServiceBus/blob/master/src/Transport/Topology/QueueAndTopicByEndpoint/NamingConventions.cs) that was in place in previous versions, where the current (not the spikes) implementation groups validation & sanitization based on path/name (queues/topics and subscriptions/rules).

Review of the 3 spikes has resulted in conclusion that none of them is solving the problem. Shortcomings are:

1. Spike #255 - `ValidationResult` passed between validation and sanitization phases becomes stale (Example: Assuming validation yielded both, characters and length were illegal, when characters are sanitized, it _could_ make path/name valid, but validation result will still indicate that it wasn't).
1. Spike #257 - while uniting validation and sanitization is a positive change, the drawback was that an end user has to now a lot of internal in order to implement a custom strategy.
1. Spike #260 - the same drawback as the second spike.

#### Proposal

Creating custom strategies (for validation and/or sanitization) is very challenging. In addition to that, the need to be able to have control over validation/sanitization of specific entity types requires a different approach. Suggested idea is to introduce another spike with the following points:

1. Modify configuration API to allow providing validation/sanitization ~~regexes~~ rules per entity type. Each "rule" is composed of 2 parts: one for validation and one for sanitization. _(Insures we're backwards compatible)_
  1. Support for v6 and v7 ~~regexes~~ rules will be provided out of the box by the transport _(v7 to be **always** the default)
1. Two sanitization strategies OOTB that take care of everything (characters _and_ length validation, sanitization, and hashing when needed).

The "strategies" we need to support:

1. [default] `ThrowOnExceptionsSanitization` with V7 rules *
1. `ValidateAndHashIfNeededStrategy` with V7 rules and MD5 hashing** by default

\* rules can be modified via configuration API. For `ThrowOnExceptionsSanitization` only validation rules modification will be acceptable and sanitization rule will be throwing an exception(?). 

\** hashing will be applied only if length is exceeding the maximum

#### Code mockups

```c#
// ThrowOnExceptionsSanitization is turned on by default

// Configuration for queue; both validation and sanitization
transport.Sanitization().
              .QueuesValidation(func<string, string>)
              .QueuesSanitization(func<string, string>); // throws because ThrowOnExceptionsSanitization doesn't perform sanitization

// Configuration for subscription; validation only; ThrowOnExceptionsSanitization is used
transport.Sanitization()
              .SubscriptionsSanitization(func<string, string>);

// Configuration for rule; sanitization only; ThrowOnExceptionsSanitization is used
transport.Sanitization()
              .RulesSanitizationExpression("rule-validation-regex");

// default v7 rules for validation & sanitization; MD5 for hashing
// rules and hashing can be custom provided
transport.Sanitization().UseStrategy<ValidateAndHashIfNeeded>()
	              .QueuesValidation(func<string, string>)
                      .QueuesSanitization(func<string, string>)
	              .HashUsing(func<string, string>);
```

**Comment by melkio** [comment]
Sorry @SeanFeldman but I don't understand the scenario covered by the last "mockup"

```c#
// default v7 rules for validation & sanitization; MD5 for hashing
// rules and hashing can be custom provided
transport.Sanitization().UseStrategy<ValidateAndHashIfNeeded>()
	              .QueuesValidation(func<string, string>)
                      .QueuesSanitization(func<string, string>)
	              .HashUsing(func<string, string>);
```

Is the user registering a strategy and then are they overriding only specific "parts"?

**Comment by SeanFeldman** [comment]
Correct @melkio. `ValidateAndHashIfNeeded` strategy is selected and one can override validation, sanitization, and hashing operation.

**Comment by melkio** [comment]
Ok, last question only to be sure to get the point.
First scenario is using the `default strategy` and user is overriding `validation` and `sanitization` for queues. Is it right?

**Comment by SeanFeldman** [comment]
Correct @melkio And since default is `ThrowOnExceptionsSanitization`, sanitation registration will not be allowed.

**Comment by SeanFeldman** [comment]
My concern is about https://github.com/Particular/NServiceBus.AzureServiceBus/pull/265#issuecomment-227048320

**Comment by SeanFeldman** [comment]
For documentation, we could add the following rules used by V6/V7 to validate and sanitise entity paths/names:

**Validation Rules, V6**
// All entities in v6 only allowed letters, numbers, periods (.), hyphens (-), and underscores (_)
Regex EntityValidationRule = new Regex(@"^[0-9A-Za-z_\.\-]+$");

**Validation Rules, V7**
// Entity segments can contain only letters, numbers, periods (.), hyphens (-), underscores (-), and slashes (/) in the middle
Regex QueueAndTopicPathValidationRule = new Regex(@"^[^\/][0-9A-Za-z_\.\-\/]+[^\/]$");

// Entity segments can contain only letters, numbers, periods (.), hyphens (-), and underscores (-)
Regex SubscriptionAndRuleNameValidationRule = new Regex(@"^[0-9A-Za-z_\.\-]+$");

**Sanitization Rules, V6**
// Remove any character other than letters, numbers, periods (.), hyphens (-), and underscores (_)
Regex EntitySanitizationRule = new Regex(@"[^a-zA-Z0-9\-._]");

**Comment by yvesgoeleven** [comment]
Updated compat & perf tests, pushed to nuget
